test/FunctionalTests/Microsoft.OData.Client.Tests/DataServiceQueryProviderTests.cs

```diff
@@ -0,0 +1,121 @@
+            var search = "Strawberry";
+            var products = dsc.CreateQuery<Product>("Products")
+                .Where(product => product.Name.Contains(search));
+
+            // Act
+            var queryComponents = sut.Translate(products.Expression);
+
+            // Assert
+            Assert.Equal(@"http://root/Products?$filter=contains(Name,'Strawberry')", queryComponents.Uri.ToString());
+        }
+
+        #endregion
+
+        #region Enumerable.Contains tests
+
+        [Fact]
+        public void TranslatesEnumerableContainsToInOperator()
+        {
+            // Arrange
+            var sut = new DataServiceQueryProvider(dsc);
+            var productNames = new[] { "Milk", "Cheese", "Donut" };
+            var products = dsc.CreateQuery<Product>("Products")
+                .Where(product => productNames.Contains(product.Name));
+
+            // Act
+            var queryComponents = sut.Translate(products.Expression);
+
+            // Assert
+            Assert.Equal(@"http://root/Products?$filter=Name in ('Milk','Cheese','Donut')", queryComponents.Uri.ToString());
+        }
+
+        [Fact]
+        public void TranslatesEnumerableContainsWithEmptyCollection()
+        {
+            // Arrange
+            var sut = new DataServiceQueryProvider(dsc);
+            var products = dsc.CreateQuery<Product>("Products")
+                .Where(product => Enumerable.Empty<string>().Contains(product.Name));
+
+            // Act
+            var queryComponents = sut.Translate(products.Expression);
+
+            // Assert
+            Assert.Equal(@"http://root/Products?$filter=Name in ()", queryComponents.Uri.ToString());
+        }
+
+        [Fact]
+        public void TranslatesEnumerableContainsWithSingleItem()
+        {
+            // Arrange
+            var sut = new DataServiceQueryProvider(dsc);
+            var products = dsc.CreateQuery<Product>("Products")
+                .Where(product => new[] { "Pancake mix" }.Contains(product.Name));
+
+            // Act
+            var queryComponents = sut.Translate(products.Expression);
+
+            // Assert
+            Assert.Equal(@"http://root/Products?$filter=Name in ('Pancake mix')", queryComponents.Uri.ToString());
+        }
+
+        [Fact]
+        public void TranslatesEnumerableContainsWithOtherClauses()
+        {
+            // Arrange
+            var sut = new DataServiceQueryProvider(dsc);
+            var products = dsc.CreateQuery<Product>("Products")
+                .Where(product => new[] { "Milk", "Flour" }.Contains(product.Name) && product.Price < 5);
+
+            // Act
+            var queryComponents = sut.Translate(products.Expression);
+
+            // Assert
+            Assert.Equal(@"http://root/Products?$filter=Name in ('Milk','Flour') and Price lt 5", queryComponents.Uri.ToString());
+        }
+
+        #endregion
+
+        [EntityType]
+        [Key(nameof(Id))]
+        private class Product

```

Comment:
>Product [](start = 22, length = 7)

to Sam's comment, can you add a collection-valued property (i.e., public string[] comments {get; set;} and then add a test for:
            var products = dsc.CreateQuery<Product>("Products")
                .Where(product => comments.Contains("buy");

and:
            var products = dsc.CreateQuery<Product>("Products")
                .Where(product => comments.Contains(product.Name);

Reply:
Oof... I'm having a hard time with this one.

Before my change, that would just throw.

> System.NotSupportedException
> The method 'Contains' is not supported.

After my change, it generates nonsensical URIs or throws more obscure exceptions. 😅

Reply:
What's the status of these cases?

Preference would be to support these, but min bar would be to raise a useful exception. I don't really want to ship generating invalid URIs...

---
In reply to: [543716880](https://github.com/OData/odata.net/pull/1868#discussion_r543716880) [](ancestors = 543716880)

Reply:
I added the suggested test cases but they simply reflect the old behavior, because I never meant to add support for these type of expressions.